src/gpio.rs

```diff
@@ -50,1205 +73,1450 @@ pub trait GpioExt {
+pub struct Input;
+/// Output mode (type state)
+pub struct Output<OTYPE>(PhantomData<OTYPE>);
+/// Alternate function (type state)
+pub struct Alternate<AF, OTYPE>(PhantomData<AF>, PhantomData<OTYPE>);
+/// Analog mode (type state)
+pub struct Analog;
 
-/// Alternate function 11 (type state)
-pub struct AF11;
+/// Push-pull output (type state)
+pub struct PushPull;
+/// Open-drain output (type state)
+pub struct OpenDrain;
 
-/// Alternate function 12 (type state)
-pub struct AF12;
+impl Readable for Input {}
+impl Readable for Output<OpenDrain> {}
+impl<OTYPE> OutputSpeed for Output<OTYPE> {}
+impl<AF, OTYPE> OutputSpeed for Alternate<AF, OTYPE> {}
+impl Active for Input {}
+impl<OTYPE> Active for Output<OTYPE> {}
+impl<AF, OTYPE> Active for Alternate<AF, OTYPE> {}
+
+/// Slew rate configuration
+pub enum Speed {
+    /// Low speed
+    Low,
+    /// Medium speed
+    Medium,
+    /// High speed
+    High,
+}
 
-/// Alternate function 13 (type state)
-pub struct AF13;
+/// Internal pull-up and pull-down resistor configuration
+pub enum Resistor {
+    /// Floating
+    Floating,
+    /// Pulled up
+    PullUp,
+    /// Pulled down
+    PullDown,
+}
 
-/// Alternate function 14 (type state)
-pub struct AF14;
+/// GPIO interrupt trigger edge selection
+pub enum Edge {
+    /// Rising edge of voltage
+    Rising,
+    /// Falling edge of voltage
+    Falling,
+    /// Rising and falling edge of voltage
+    RisingFalling,
+}
 
-/// Alternate function 15 (type state)
-pub struct AF15;
+/// Generic pin
+pub struct Pin<GPIO, INDEX, MODE> {
+    gpio: GPIO,
+    index: INDEX,
+    _mode: PhantomData<MODE>,
+}
 
-macro_rules! gpio {
-    ([
-        $({
-            GPIO: $GPIOX:ident,
-            gpio: $gpiox:ident,
-            gpio_mapped: $gpioy:ident,
-            gpio_mapped_ioenr: $iopxenr:ident,
-            gpio_mapped_iorst: $iopxrst:ident,
-            partially_erased_pin: $PXx:ident,
-            pins: [
-                $($PXi:ident: (
-                    $pxi:ident, $i:expr, $MODE:ty, $moderi:ident, $AFR:ident, $afri:ident,
-                    $bsi:ident, $bri:ident, $odri:ident, $idri:ident, $pupdri:ident, $oti:ident,
-                    { $( $AFi:ty: ($into_afi:ident, $afi:ident), )* },
-                ),)+
-            ],
-        },)+
-    ]) => {
-        $( use crate::pac::$GPIOX; )+
+/// Fully erased pin
+///
+/// This moves the pin type information to be known
+/// at runtime, and erases the specific compile time type of the GPIO.
+/// It does only matter, that it is a GPIO pin with a specific MODE.
+///
+/// See [examples/gpio_erased.rs] as an example.
+///
+/// [examples/gpio_erased.rs]: https://github.com/stm32-rs/stm32f3xx-hal/blob/v0.6.0/examples/gpio_erased.rs
+pub type PXx<MODE> = Pin<Gpiox, Ux, MODE>;
+
+macro_rules! modify_at {
+    ($xr:expr, $bits:expr, $i:expr, $val:expr) => {
+        $xr.modify(|r, w| {
+            w.bits(r.bits() & !(u32::MAX >> 32 - $bits << $bits * $i) | $val << $bits * $i)
```

Comment:
I had to take some time to understand it, but i'd suggest to write the macro like this and comment it a bit more, so that it is easier to parse. Without the knowledge of the operator evaluation order, this is hard to parse.

```suggestion
            let mask = !(u32::MAX >> (32 - $bits) << ($bits * $i));
            let value = $val << $bits * $i;
            w.bits(r.bits() & mask | value)
```